core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionClassLoader.java

```diff
@@ -76,22 +82,36 @@ protected Class<?> findClass(String name) throws ClassNotFoundException {
     }
 
     protected Class<?> findClass(String name, boolean checkGlobal) throws ClassNotFoundException {
+        // Don't let extensions load classes from Geyser or minecraft packages without a warning
         if (name.startsWith("org.geysermc.geyser.") || name.startsWith("net.minecraft.")) {
-            throw new ClassNotFoundException(name);
+            if (!warned) {
+                GeyserImpl.getInstance().getLogger().warning("Extension " + this.description.name() + " loads class " + name + " from the Geyser package." +
+                        "This can change at any time and break the extension, additionally to causing unexpected behaviour!");
+                warned = true;
+            }
```

Comment:
I feel like it may be worth making this a system property or an option in the extension.yml that has to be explicitly defined, like a `allow-internal-code-references` option. Feel free to suggest a better name, but I'm not super big on allowing for unsafe internal references by default.

Reply:
Implemented in https://github.com/GeyserMC/Geyser/pull/3946/commits/0052c6456ef4d6f2a1eb82b9d9ddc24b1a787436, named it allowInternalCodeReferences. If not set, defaults to false.

slight issue atm: That option/warning does not appear currently for extensions such as MagicMenu or MCXboxBroadcast which all use internals. Unfortunately, i don't know enough about classloading etc to diagnose why it's not properly called to load internals; according to rtm, maybe it's due to the parent classloader loading the classes.. idek